### PR TITLE
Fixed noise_pred_text referenced before assignment.

### PR DESCRIPTION
**Pull request by LagPixelLOL**
Fixed local variable noise_pred_text referenced before assignment when using PAG with guidance scale and guidance rescale at the same time.

@yiyixuxu @asomoza

I'm not sure if allowing guidance rescale with PAG + guidance scale is actually a thing, but from my tests it does work with PAG + guidance scale. If it's not allowed then it can be fixed with adding a condition or raising an error when all PAG scale, guidance scale, and guidance rescale are passed at the same time.

Before this fix it currently does raise an error when passing all 3 at the same time, but it's an unbounded local variable error, which is extremely confusing.

### CODE REVIEW

**Comment by LagPixelLOL** [comment]
@sayakpaul

**Comment by sayakpaul** [comment]
Thanks for your PR, could you provide a reproducible snippet that raises the error?

**Comment by sayakpaul** [comment]
I am also going to cc @sunovivid (PAG first author) to brainstorm if allowing guidance rescale with PAG + guidance scale makes sense.

**Comment by LagPixelLOL** [comment]
```py
import torch
import diffusers

pipeline = diffusers.AutoPipelineForText2Image.from_pretrained("stabilityai/stable-diffusion-xl-base-1.0", torch_dtype=torch.bfloat16, enable_pag=True)
pipeline.to("cuda")
pipeline(
    prompt="test",
    guidance_scale=5,
    guidance_rescale=0.5,
    pag_scale=3,
    num_inference_steps=35,
)
```
produces the below error
```
Traceback (most recent call last):
  File "/root/differr.py", line 6, in <module>
    pipeline(
  File "/usr/local/lib/python3.12/site-packages/torch/utils/_contextlib.py", line 116, in decorate_context
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/diffusers/pipelines/pag/pipeline_pag_sd_xl.py", line 1250, in __call__
    noise_pred = rescale_noise_cfg(noise_pred, noise_pred_text, guidance_rescale=self.guidance_rescale)
                                               ^^^^^^^^^^^^^^^
UnboundLocalError: cannot access local variable 'noise_pred_text' where it is not associated with a value
```

**Comment by yiyixuxu** [comment]
can you run `make style` and `make fix-copies`?

**Comment by sunovivid** [comment]
> I am also going to cc @sunovivid (PAG first author) to brainstorm if allowing guidance rescale with PAG + guidance scale makes sense.

It looks reasonable to use guidance rescale with PAG + guidance scale. Noise rescale matches the scale of the 'guided' prediction to the original text-conditioned prediction to cut off extra scale from guidance extrapolation. So rescaling PAG + CFG guided prediction to the original prediction looks reasonable. After the cut-off, the noise rescale function mixes the rescaled noise with the original 'text' prediction. It seems also reasonable because our main component is 'text' prediction, both in CFG and PAG. I think it is a very good fix!

**Comment by LagPixelLOL** [comment]
The style is fixed now.

File: src/diffusers/pipelines/pag/pag_utils.py [hunk 1]
```diff
@@ -122,7 +123,7 @@ def _apply_perturbed_attention_guidance(self, noise_pred, do_classifier_free_gui
         else:
             noise_pred_text, noise_pred_perturb = noise_pred.chunk(2)
             noise_pred = noise_pred_text + pag_scale * (noise_pred_text - noise_pred_perturb)
-        return noise_pred
+        return noise_pred, noise_pred_text
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/pag/pag_utils.py:126]: seems like we broke the tests; I looked up the import of PAGMixin on github, not many but some so let's try to not making a breaking change here

we can add a argument `return_pred_text`; default to `False` but set `True` when calling from our pipelines.

**Comment by LagPixelLOL** [comment]
Should be fixed now.

**Comment by yiyixuxu** [comment]
@LagPixelLOL merged! thanks for your PR!

**Pull request merged**